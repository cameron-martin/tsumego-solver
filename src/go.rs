mod benson;
mod bit_board;
mod sgf_conversion;
pub use bit_board::{BitBoard, BoardPosition};
use std::collections::hash_map::DefaultHasher;
use std::fmt;
use std::fmt::Debug;
use std::fmt::Formatter;
use std::fmt::{Display, Write};
use std::hash::{Hash, Hasher};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Move {
    PassTwice,
    PassOnce,
    Place(BoardPosition),
}

impl Display for Move {
    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
        match self {
            Move::PassOnce => f.write_str("Pass (once)"),
            Move::PassTwice => f.write_str("Pass (twice)"),
            Move::Place(position) => f.write_fmt(format_args!("{}", position)),
        }
    }
}

#[derive(Copy, Clone, Debug, PartialEq)]
pub enum BoardCell {
    Empty,
    OutOfBounds,
    Occupied(GoPlayer),
}

#[derive(Copy, Clone, Debug, PartialEq)]
pub enum GoPlayer {
    Black,
    White,
}

impl GoPlayer {
    pub fn flip(self) -> GoPlayer {
        match self {
            GoPlayer::Black => GoPlayer::White,
            GoPlayer::White => GoPlayer::Black,
        }
    }

    pub fn both() -> impl Iterator<Item = &'static GoPlayer> {
        [GoPlayer::Black, GoPlayer::White].iter()
    }
}

impl Display for GoPlayer {
    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
        match self {
            GoPlayer::Black => f.write_str("black"),
            GoPlayer::White => f.write_str("white"),
        }
    }
}

// Being set in both black and white denotes "out of bounds"
#[derive(PartialEq, Clone, Copy, Debug, Hash)]
pub struct GoBoard {
    white: BitBoard,
    black: BitBoard,
}

impl Display for GoBoard {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for j in 0..BitBoard::height() {
            for i in 0..BitBoard::width() {
                if i != 0 {
                    f.write_char(' ')?;
                }

                f.write_char(match self.get_cell(BoardPosition::new(i, j)) {
                    BoardCell::Empty => '.',
                    BoardCell::OutOfBounds => '_',
                    BoardCell::Occupied(GoPlayer::White) => 'w',
                    BoardCell::Occupied(GoPlayer::Black) => 'b',
                })?;
            }
            f.write_str("\n")?;
        }

        Ok(())
    }
}

impl GoBoard {
    fn empty() -> GoBoard {
        GoBoard {
            white: BitBoard::empty(),
            black: BitBoard::empty(),
        }
    }

    pub fn new(black: BitBoard, white: BitBoard, out_of_bounds: BitBoard) -> GoBoard {
        GoBoard {
            white: white | out_of_bounds,
            black: black | out_of_bounds,
        }
    }

    /// Empty cells, including out of bounds
    fn empty_cells(&self) -> BitBoard {
        !(self.white ^ self.black)
    }

    pub fn out_of_bounds(&self) -> BitBoard {
        self.white & self.black
    }

    fn set_cell(&mut self, position: BoardPosition, cell: BoardCell) {
        let mask = BitBoard::singleton(position);

        match cell {
            BoardCell::Empty => {
                self.black = self.black & !mask;
                self.white = self.white & !mask;
            }
            BoardCell::Occupied(GoPlayer::Black) => {
                self.black = self.black | mask;
                self.white = self.white & !mask;
            }
            BoardCell::Occupied(GoPlayer::White) => {
                self.black = self.black & !mask;
                self.white = self.white | mask;
            }
            BoardCell::OutOfBounds => {
                self.black = self.black | mask;
                self.white = self.white | mask;
            }
        }
    }

    fn get_cell(&self, position: BoardPosition) -> BoardCell {
        let mask = BitBoard::singleton(position);

        if !((mask & self.out_of_bounds()).is_empty()) {
            return BoardCell::OutOfBounds;
        }

        if !((mask & self.white).is_empty()) {
            return BoardCell::Occupied(GoPlayer::White);
        }

        if !((mask & self.black).is_empty()) {
            return BoardCell::Occupied(GoPlayer::Black);
        }

        BoardCell::Empty
    }

    pub fn get_bitboard_for_player(&self, player: GoPlayer) -> BitBoard {
        match player {
            GoPlayer::Black => (self.black & !self.white),
            GoPlayer::White => (self.white & !self.black),
        }
    }

    fn set_bitboard_for_player(&mut self, player: GoPlayer, board: BitBoard) {
        match player {
            GoPlayer::Black => self.black = board | self.out_of_bounds(),
            GoPlayer::White => self.white = board | self.out_of_bounds(),
        }
    }

    fn get_bitboard_at_position(&self, position: BoardPosition) -> BitBoard {
        let mask = BitBoard::singleton(position);
        let white = self.get_bitboard_for_player(GoPlayer::White);
        let black = self.get_bitboard_for_player(GoPlayer::Black);

        if !((mask & white).is_empty()) {
            return white;
        }

        if !((mask & black).is_empty()) {
            return black;
        }

        panic!("No board at this position")
    }

    fn group_has_liberties(&self, position: BoardPosition) -> bool {
        let mask = self.get_bitboard_at_position(position);

        let group = BitBoard::singleton(position).flood_fill(mask);

        !(group.expand_one() & self.empty_cells()).is_empty()
    }

    fn get_alive_groups_for_player(&self, player: GoPlayer) -> BitBoard {
        let bitboard = self.get_bitboard_for_player(player);

        (self.empty_cells().expand_one() & bitboard).flood_fill(bitboard)
    }

    fn remove_dead_groups_for_player(&mut self, player: GoPlayer) {
        let stones_with_liberties = self.get_alive_groups_for_player(player);

        self.set_bitboard_for_player(player, stones_with_liberties);
    }

    pub fn has_dead_groups(&self) -> bool {
        GoPlayer::both().any(|&player| {
            let alive_groups = self.get_alive_groups_for_player(player);

            self.get_bitboard_for_player(player) != alive_groups
        })
    }

    fn is_out_of_bounds(&self, position: BoardPosition) -> bool {
        !(BitBoard::singleton(position) & self.out_of_bounds()).is_empty()
    }

    fn set_out_of_bounds(&mut self, out_of_bounds: BitBoard) {
        let prev_out_of_bounds = self.out_of_bounds();
        self.white = (self.white & !prev_out_of_bounds) | out_of_bounds;
        self.black = (self.black & !prev_out_of_bounds) | out_of_bounds;
    }

    pub fn stable_hash(&self) -> u64 {
        let mut hasher = DefaultHasher::new();
        self.hash(&mut hasher);

        hasher.finish()
    }
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct GoGame {
    ko_violations: BitBoard,
    board: GoBoard,
    pub current_player: GoPlayer,
    pub last_move_pass: bool,
}

#[derive(Debug, PartialEq)]
pub enum MoveError {
    Occupied,
    OutOfBounds,
    OutOfTurn,
    Suicidal,
    Ko,
}

impl GoGame {
    pub fn empty() -> GoGame {
        GoGame {
            board: GoBoard::empty(),
            ko_violations: BitBoard::empty(),
            current_player: GoPlayer::Black,
            last_move_pass: false,
        }
    }

    pub fn from_board(board: GoBoard, current_player: GoPlayer) -> GoGame {
        GoGame {
            board,
            ko_violations: BitBoard::empty(),
            current_player,
            last_move_pass: false,
        }
    }

    pub fn get_board(&self) -> GoBoard {
        self.board
    }

    fn get_cell(&self, position: BoardPosition) -> BoardCell {
        self.get_board().get_cell(position)
    }

    fn is_out_of_bounds(&self, position: BoardPosition) -> bool {
        self.get_board().is_out_of_bounds(position)
    }

    pub fn play_move_for_player(
        &self,
        go_move: Move,
        player: GoPlayer,
    ) -> Result<GoGame, MoveError> {
        if self.current_player != player {
            return Err(MoveError::OutOfTurn);
        }

        self.play_move(go_move)
    }

    pub fn play_move(&self, go_move: Move) -> Result<GoGame, MoveError> {
        match go_move {
            Move::Place(position) => self.play_placing_move(position),
            Move::PassOnce => Ok(self.pass()),
            Move::PassTwice => Ok(self.pass()),
        }
    }

    pub fn play_placing_move(&self, position: BoardPosition) -> Result<GoGame, MoveError> {
        if self.get_cell(position) != BoardCell::Empty {
            return Err(MoveError::Occupied);
        }

        if self.is_out_of_bounds(position) {
            return Err(MoveError::OutOfBounds);
        }

        let next_player = self.current_player.flip();

        let mut new_board = self.get_board();
        new_board.set_cell(position, BoardCell::Occupied(self.current_player));

        // Remove dead groups owned by other player
        new_board.remove_dead_groups_for_player(next_player);

        // Evaluate suicide
        if !new_board.group_has_liberties(position) {
            return Err(MoveError::Suicidal);
        }

        // Evaluate ko
        if self.ko_violations.is_set(position) {
            return Err(MoveError::Ko);
        }

        let ko_violations = if (BitBoard::singleton(position).immediate_exterior()
            & self
                .get_board()
                .get_bitboard_for_player(self.current_player))
        .is_empty()
        {
            (self.board.get_bitboard_for_player(next_player)
                & !new_board.get_bitboard_for_player(next_player))
            .singletons()
        } else {
            BitBoard::empty()
        };

        Ok(GoGame {
            ko_violations,
            board: new_board,
            current_player: next_player,
            last_move_pass: false,
        })
    }

    pub fn pass(&self) -> GoGame {
        GoGame {
            board: self.board,
            ko_violations: BitBoard::empty(),
            current_player: self.current_player.flip(),
            last_move_pass: true,
        }
    }

    pub fn generate_moves(&self) -> Vec<(GoGame, Move)> {
        let mut games = Vec::new();

        let board = self.get_board();

        for position in (!(board.white | board.black)).positions() {
            if let Ok(game) = self.play_placing_move(position) {
                games.push((game, Move::Place(position)));
            }
        }

        games
    }

    pub fn generate_moves_including_pass(&self) -> Vec<(GoGame, Move)> {
        let mut games = self.generate_moves();

        games.push((
            self.pass(),
            if self.last_move_pass {
                Move::PassTwice
            } else {
                Move::PassOnce
            },
        ));

        games
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn can_add_stone() {
        let game = GoGame::empty();
        let game = game
            .play_move_for_player(Move::Place(BoardPosition::new(0, 0)), GoPlayer::Black)
            .unwrap();

        assert_eq!(
            game.get_cell(BoardPosition::new(0, 0)),
            BoardCell::Occupied(GoPlayer::Black)
        );
    }

    #[test]
    fn previous_board_is_not_mutated() {
        let old_game = GoGame::empty();
        let new_game = old_game
            .play_move_for_player(Move::Place(BoardPosition::new(0, 0)), GoPlayer::Black)
            .unwrap();

        assert_ne!(
            old_game.get_cell(BoardPosition::new(0, 0)),
            new_game.get_cell(BoardPosition::new(0, 0))
        );
    }

    #[test]
    fn current_player_starts_as_black() {
        let game = GoGame::empty();

        assert_eq!(game.current_player, GoPlayer::Black);
    }

    #[test]
    fn player_advances_when_playing_move() {
        let game = GoGame::empty()
            .play_placing_move(BoardPosition::new(0, 0))
            .unwrap();

        assert_eq!(game.current_player, GoPlayer::White);
    }

    #[test]
    fn cannot_play_move_out_of_turn() {
        let result = GoGame::empty()
            .play_move_for_player(Move::Place(BoardPosition::new(0, 0)), GoPlayer::White);

        assert_eq!(result, Err(MoveError::OutOfTurn));
    }

    #[test]
    fn cannot_play_in_occupied_space() {
        let game = GoGame::empty()
            .play_placing_move(BoardPosition::new(0, 0))
            .unwrap();
        let result = game.play_placing_move(BoardPosition::new(0, 0));

        assert_eq!(result, Err(MoveError::Occupied));
    }

    #[test]
    fn single_groups_are_captured() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/single_groups_are_captured.sgf"));

        assert_eq!(game.get_cell(BoardPosition::new(0, 0)), BoardCell::Empty);
    }

    #[test]
    fn complex_groups_are_captured() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/complex_capture.sgf"));
        let game = game.play_placing_move(BoardPosition::new(11, 6)).unwrap();

        assert_eq!(
            format!("{}", game.get_board()),
            ". . . . . . . . . . . . . . . .\n\
             . b . b . b b w w . w . . . . .\n\
             . . . . . . . b w w . w . w . .\n\
             . . . b . b b . w . . w b . . .\n\
             . . . . b . . . w w . w . . . .\n\
             . . . b . . w . w . . . w . . .\n\
             . . . . . . . . . w w w . . . .\n\
             . . . . . . . . . . . . . . . .\n"
        );
    }

    #[test]
    fn capturing_has_precedence_over_suicide() {
        let game = GoGame::from_sgf(include_str!(
            "test_sgfs/capturing_has_precedence_over_suicide.sgf"
        ));

        assert_eq!(game.get_cell(BoardPosition::new(1, 0)), BoardCell::Empty);
    }

    #[test]
    fn cannot_commit_suicide() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/cannot_commit_suicide.sgf"));
        let result = game.play_placing_move(BoardPosition::new(0, 0));

        assert_eq!(result, Err(MoveError::Suicidal));
    }

    #[test]
    fn ko_rule_simple() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/ko_rule_simple.sgf"));
        let result = game.play_placing_move(BoardPosition::new(2, 2));

        assert_eq!(result, Err(MoveError::Ko));
    }

    #[test]
    fn capture_two_recapture_one_not_ko_violation() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/capture_two_recapture_one.sgf"));

        game.play_placing_move(BoardPosition::new(3, 2)).unwrap();
    }

    #[test]
    fn capturing_single_and_joining_group_does_not_trigger_ko() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/capture_single_join_group.sgf"));

        let result = game.play_placing_move(BoardPosition::new(2, 1));

        assert!(result.is_ok());
    }

    #[test]
    fn out_of_bounds_moves_are_not_generated() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple1.sgf"));
        let moves = game.generate_moves();

        assert_eq!(moves.len(), 5);
    }

    #[test]
    fn pass_sets_last_move_pass() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/ko_rule_simple.sgf"));
        let new_game = game.pass();

        assert!(new_game.last_move_pass);
    }

    #[test]
    fn move_clears_last_move_pass() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/ko_rule_simple.sgf"));
        let game = game.pass();
        let game = game.play_placing_move(BoardPosition::new(13, 7)).unwrap();

        assert!(!game.last_move_pass);
    }

    #[test]
    fn pass_advances_player() {
        let game = GoGame::from_sgf(include_str!("test_sgfs/ko_rule_simple.sgf"));
        let new_game = game.pass();

        assert_ne!(game.current_player, new_game.current_player);
    }

    #[test]
    fn has_dead_groups_black() {
        let mut game = GoBoard::empty();
        game.set_cell(
            BoardPosition::new(0, 0),
            BoardCell::Occupied(GoPlayer::Black),
        );
        game.set_cell(
            BoardPosition::new(0, 1),
            BoardCell::Occupied(GoPlayer::White),
        );
        game.set_cell(
            BoardPosition::new(1, 0),
            BoardCell::Occupied(GoPlayer::White),
        );

        assert!(game.has_dead_groups());
    }

    #[test]
    fn has_dead_groups_white() {
        let mut game = GoBoard::empty();
        game.set_cell(
            BoardPosition::new(0, 0),
            BoardCell::Occupied(GoPlayer::White),
        );
        game.set_cell(
            BoardPosition::new(0, 1),
            BoardCell::Occupied(GoPlayer::Black),
        );
        game.set_cell(
            BoardPosition::new(1, 0),
            BoardCell::Occupied(GoPlayer::Black),
        );

        assert!(game.has_dead_groups());
    }

    #[test]
    fn has_dead_groups_false() {
        let mut game = GoBoard::empty();
        game.set_cell(
            BoardPosition::new(0, 1),
            BoardCell::Occupied(GoPlayer::White),
        );
        game.set_cell(
            BoardPosition::new(1, 0),
            BoardCell::Occupied(GoPlayer::White),
        );

        assert!(!game.has_dead_groups());
    }

    #[test]
    fn hashing_is_stable() {
        assert_eq!(GoBoard::empty().stable_hash(), 13284472273662876477);
    }
}
