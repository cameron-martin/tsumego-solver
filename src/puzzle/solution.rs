use crate::go::Move;

pub struct Solution {
    pub won: bool,
    pub principle_variation: Vec<Move>,
}
