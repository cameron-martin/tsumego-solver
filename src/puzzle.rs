mod profiler;
mod solution;
mod terminal_detection;

use crate::go::{BoardPosition, GoGame, GoPlayer, Move};
pub use profiler::{NoProfile, Profile, Profiler};
use solution::Solution;
use std::{
    collections::HashSet,
    iter,
    time::{Duration, Instant},
};

trait AbortController {
    fn should_abort(&self) -> bool;
}

struct NoAbortController;

impl AbortController for NoAbortController {
    fn should_abort(&self) -> bool {
        false
    }
}

struct TimeoutAbortController {
    timeout_at: Instant,
}

impl AbortController for TimeoutAbortController {
    fn should_abort(&self) -> bool {
        Instant::now() >= self.timeout_at
    }
}

impl TimeoutAbortController {
    fn duration(duration: Duration) -> Self {
        TimeoutAbortController {
            timeout_at: Instant::now() + duration,
        }
    }
}

pub struct Puzzle<P: Profiler> {
    game: GoGame,
    player: GoPlayer,
    attacker: GoPlayer,
    pub profiler: P,
}

impl<P: Profiler> Puzzle<P> {
    pub fn new(game: GoGame) -> Puzzle<P> {
        let attacker = if !(game.get_board().out_of_bounds().expand_one()
            & game.get_board().get_bitboard_for_player(GoPlayer::White))
        .is_empty()
        {
            GoPlayer::White
        } else {
            GoPlayer::Black
        };

        let player = game.current_player;

        Puzzle {
            game,
            player,
            attacker,
            profiler: P::new(),
        }
    }

    pub fn from_sgf(sgf_string: &str) -> Puzzle<P> {
        Self::new(GoGame::from_sgf(sgf_string))
    }

    pub fn solve(&mut self) -> Solution {
        self.solve_with_controller(&NoAbortController).unwrap()
    }

    pub fn solve_with_timeout(&mut self, timeout: Duration) -> Option<Solution> {
        self.solve_with_controller(&TimeoutAbortController::duration(timeout))
    }

    fn solve_with_controller<C: AbortController>(&mut self, controller: &C) -> Option<Solution> {
        let mut parents = HashSet::new();
        parents.insert(self.game);

        let mut max_depth: u8 = 1;

        loop {
            let mut variations = iter::repeat(Move::Place(BoardPosition::new(0, 0)))
                .take((max_depth as usize * (max_depth as usize + 1)) / 2)
                .collect();

            let result = self.negamax(
                self.game,
                -std::i8::MAX,
                std::i8::MAX,
                0,
                max_depth,
                1,
                controller,
                &mut parents,
                &mut variations,
                0,
            )?;

            if result != 0 {
                variations.truncate(max_depth as usize);

                return Some(Solution {
                    won: result > 0,
                    principle_variation: variations,
                });
            }

            max_depth += 1;
            self.profiler.move_down();
        }
    }

    fn negamax<C: AbortController>(
        &mut self,
        node: GoGame,
        alpha: i8,
        beta: i8,
        depth: u8,
        max_depth: u8,
        is_maximising_player: i8,
        controller: &C,
        parents: &mut HashSet<GoGame>,
        variations: &mut Vec<Move>,
        variations_index: usize,
    ) -> Option<i8> {
        if controller.should_abort() {
            return None;
        }

        self.profiler.visit_node();

        if let Some(value) = terminal_detection::is_terminal(node, self.player, self.attacker) {
            return Some(
                is_maximising_player
                    * if value {
                        std::i8::MAX - depth as i8
                    } else {
                        -(std::i8::MAX - depth as i8)
                    },
            );
        }

        if depth == max_depth {
            return Some(0);
        }

        let mut alpha = alpha;
        let this_variation_size = (max_depth - depth) as usize;
        let child_variation_size = this_variation_size - 1;
        let child_variations_index = variations_index + this_variation_size;

        let mut m = -std::i8::MAX;
        for (child, go_move) in node.generate_moves_including_pass() {
            if parents.contains(&child) {
                continue;
            }
            parents.insert(child);

            let t = -self.negamax(
                child,
                -beta,
                -alpha,
                depth + 1,
                max_depth,
                -is_maximising_player,
                controller,
                parents,
                variations,
                child_variations_index,
            )?;
            if t > m {
                m = t;
            }
            parents.remove(&child);
            if m >= beta {
                break;
            }

            if m > alpha {
                alpha = m;

                // Update principal variation
                variations[variations_index] = go_move;

                let (dst_arr, src_arr) = variations
                    [variations_index + 1..child_variations_index + child_variation_size]
                    .split_at_mut(child_variation_size);
                for (dst, src) in dst_arr.iter_mut().zip(src_arr) {
                    *dst = *src;
                }
            }
        }

        return Some(m);
    }

    // fn root_node(&self) -> AndOrNode {
    //     self.tree[self.root_id]
    // }

    // pub fn first_move(&self) -> Move {
    //     *self
    //         .tree
    //         .edges(self.root_id)
    //         .find(|edge| self.tree[edge.target()].is_proved())
    //         .unwrap()
    //         .weight()
    // }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::go::{BoardPosition, GoGame};
    use insta::{assert_debug_snapshot, assert_display_snapshot, assert_snapshot};
    use profiler::Profile;
    use std::borrow::Borrow;

    fn show_principle_variation<T: Profiler>(puzzle: &Puzzle<T>, solution: &Solution) -> String {
        let games = solution
            .principle_variation
            .iter()
            .scan(puzzle.game, |state, &go_move| {
                *state = state.play_move(go_move).unwrap();

                Some((*state, go_move))
            });

        let mut output = String::new();
        output.push_str(format!("{}\n\n", puzzle.game.get_board()).borrow());
        for (game, go_move) in games {
            output.push_str(
                format!(
                    "{}: {}\n{}\n\n",
                    game.current_player.flip(),
                    go_move,
                    game.get_board()
                )
                .borrow(),
            );
        }

        output
    }

    #[test]
    fn true_simple1() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple1.sgf"));

        let mut puzzle = Puzzle::<Profile>::new(tsumego);

        let solution = puzzle.solve();

        assert!(solution.won);
        assert_display_snapshot!(puzzle.profiler.visited_nodes, @"644");
        assert_display_snapshot!(puzzle.profiler.max_depth, @"5");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    #[test]
    fn true_simple2() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple2.sgf"));

        let mut puzzle = Puzzle::<Profile>::new(tsumego);

        let solution = puzzle.solve();

        assert!(solution.won);
        assert_display_snapshot!(puzzle.profiler.visited_nodes, @"7295");
        assert_display_snapshot!(puzzle.profiler.max_depth, @"9");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    #[test]
    fn true_simple3() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple3.sgf"));

        let mut puzzle = Puzzle::<Profile>::new(tsumego);

        let solution = puzzle.solve();

        assert!(solution.won);
        assert_display_snapshot!(puzzle.profiler.visited_nodes, @"2946");
        assert_display_snapshot!(puzzle.profiler.max_depth, @"8");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    #[test]
    fn true_simple4() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_simple4.sgf"));

        let mut puzzle = Puzzle::<Profile>::new(tsumego);

        let solution = puzzle.solve();

        assert!(solution.won);
        assert_display_snapshot!(puzzle.profiler.visited_nodes, @"31143");
        assert_display_snapshot!(puzzle.profiler.max_depth, @"7");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    // #[test]
    // fn true_medium1() {
    //     let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_medium1.sgf"));

    //     let mut puzzle = Puzzle::<Profile>::new(tsumego);

    //     let solution = puzzle.solve();

    //     assert!(solution.won);
    //     // assert_eq!(puzzle.first_move(), Move::Place(BoardPosition::new(14, 2)));
    //     assert_display_snapshot!(puzzle.profiler.visited_nodes, @"345725");
    //     assert_display_snapshot!(puzzle.profiler.max_depth, @"29");
    // }

    #[test]
    fn true_ultrasimple1() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_ultrasimple1.sgf"));

        let mut puzzle = Puzzle::<Profile>::new(tsumego);

        let solution = puzzle.solve();

        assert!(solution.won);
        assert_display_snapshot!(puzzle.profiler.visited_nodes, @"5");
        assert_display_snapshot!(puzzle.profiler.max_depth, @"1");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }

    #[test]
    fn true_ultrasimple2() {
        let tsumego = GoGame::from_sgf(include_str!("test_sgfs/puzzles/true_ultrasimple2.sgf"));

        let mut puzzle = Puzzle::<Profile>::new(tsumego);

        let solution = puzzle.solve();

        assert!(solution.won);
        assert_display_snapshot!(puzzle.profiler.visited_nodes, @"2170");
        assert_display_snapshot!(puzzle.profiler.max_depth, @"8");
        assert_snapshot!(show_principle_variation(&puzzle, &solution));
    }
}
