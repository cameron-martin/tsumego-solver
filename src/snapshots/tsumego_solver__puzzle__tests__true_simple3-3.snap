---
source: src/puzzle.rs
expression: "show_principle_variation(&puzzle, &solution)"
---
_ _ b w . . . w b _ _ _ _ _ _ _
_ _ b w w w w w b _ _ _ _ _ _ _
_ _ b b b b b b b _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


black: (5, 0)
_ _ b w . b . w b _ _ _ _ _ _ _
_ _ b w w w w w b _ _ _ _ _ _ _
_ _ b b b b b b b _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


white: Pass (once)
_ _ b w . b . w b _ _ _ _ _ _ _
_ _ b w w w w w b _ _ _ _ _ _ _
_ _ b b b b b b b _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


black: (4, 0)
_ _ b w b b . w b _ _ _ _ _ _ _
_ _ b w w w w w b _ _ _ _ _ _ _
_ _ b b b b b b b _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


white: (6, 0)
_ _ b w . . w w b _ _ _ _ _ _ _
_ _ b w w w w w b _ _ _ _ _ _ _
_ _ b b b b b b b _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


black: Pass (once)
_ _ b w . . w w b _ _ _ _ _ _ _
_ _ b w w w w w b _ _ _ _ _ _ _
_ _ b b b b b b b _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


white: (4, 0)
_ _ b w w . w w b _ _ _ _ _ _ _
_ _ b w w w w w b _ _ _ _ _ _ _
_ _ b b b b b b b _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


black: Pass (once)
_ _ b w w . w w b _ _ _ _ _ _ _
_ _ b w w w w w b _ _ _ _ _ _ _
_ _ b b b b b b b _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _


white: Pass (twice)
_ _ b w w . w w b _ _ _ _ _ _ _
_ _ b w w w w w b _ _ _ _ _ _ _
_ _ b b b b b b b _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _



